### PR TITLE
fix: fixes for v1.14.0

### PR DESCRIPTION
**Pull request by dranikpg**
1. Stop writing to the replication ring_buffer
2. Stop allocating in TopKeys 
3. Tighter CHECKs around tx execution.

### CODE REVIEW

File: src/server/journal/journal_slice.cc [hunk 1]
```diff
@@ -156,7 +156,7 @@ void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
     FiberAtomicGuard fg;
     // GetTail gives a pointer to a new tail entry in the buffer, possibly overriding the last entry
     // if the buffer is full.
-    item = ring_buffer_->GetTail(true);
+    item = &dummy;
```
Comment (adiholden) [inline src/server/journal/journal_slice.cc:159]: removing the use of ring_buffer_ from replication?

File: src/server/transaction.cc [hunk 2]
```diff
@@ -1396,10 +1394,13 @@ inline uint32_t Transaction::DecreaseRunCnt() {
   // The fact that run_ec_.notify() does release operation is not enough, because
   // WaitForCallbacks might skip reading run_ec_ if run_count_ is already 0.
   uint32_t res = run_count_.fetch_sub(1, memory_order_release);
+
+  CHECK_GE(res, 1u) << unique_shard_cnt_ << " " << unique_shard_id_ << " " << cid_->name() << " "
+                    << use_count_.load(memory_order_relaxed) << " " << coordinator_state_;
```
Comment (romange) [inline src/server/transaction.cc:1399]: cast `coordinator_state_` to int

Comment (dranikpg) [reply src/server/transaction.cc:1399]: Good catch

File: src/server/journal/journal_slice.cc (hunk 1 above)
Comment (dranikpg) [reply src/server/journal/journal_slice.cc:159]: 🤷🏻‍♂️ 

We seem to be okay with it https://github.com/dragonflydb/dragonfly/pull/2466

File: src/server/transaction.cc [hunk 3]
```diff
@@ -493,9 +493,7 @@ bool Transaction::RunInShard(EngineShard* shard, bool txq_ooo) {
   unsigned idx = SidToId(shard->shard_id());
   auto& sd = shard_data_[idx];
 
-  bool prev_armed = sd.is_armed.load(memory_order_relaxed);
-  DCHECK(prev_armed);
-  sd.is_armed.store(false, memory_order_relaxed);
+  CHECK(sd.is_armed.exchange(false, memory_order_relaxed));
```
Comment (romange) [inline src/server/transaction.cc:496]: Please also add `CHECK_GT(run_count_.load(acquire), 0)` - lets see if we catch it earlier.

Comment (dranikpg) [reply src/server/transaction.cc:496]: maybe relaxed? If we change acquisition semantics we might change the code

**Comment by dranikpg** [comment]
It's bad CI doesn't run for this branch 👀

**Pull request merged**